### ISSUE TITLE
Vector type coercion fails at times

### ISSUE THREAD

**Issue by aarvay** [ziglang/zig#6864]
```zig
const Vec2 = std.meta.Vector(2, u32);
const Strukt = struct { v: Vec2 };

test "passes" {
    const s: Strukt = .{ .v = .{ 1, 1 } };
    const v = s.v;
}

const s2: Strukt = .{ .v = .{ 1, 1 } };
test "fails" {
    const v = s2.v;
}
```
fails with
```sh
error: expected type '@Vector(2, u32)', found 'struct:270:29'
const s2: Strukt = .{ .v = .{ 1, 1 } };
                    ^
./src/test.zig:270:29: note: struct:270:29 declared here
const s2: Strukt = .{ .v = .{ 1, 1 } };
                            ^
./src/test.zig:272:15: note: referenced here
    const v = s2.v;
```

Why would the scope affect type coercion here?

**Comment by dbandstra** [comment]
I think some type inference features are broken or not implemented at global scope. I suspect that in the global declaration, Strukt is only being applied shallowly to the anonymous literal on the right hand side.

**Comment by mikdusan** [comment]
with latest compiler:
- compiler no longer errors
- appears this was fixed between 0.9.1 and 0.10.0

<details>
<summary>repro.zig</summary>

```zig
const std = @import("std");

const Vec2 = std.meta.Vector(2, u32);
const Strukt = struct { v: Vec2 };

test "passes" {
    const s: Strukt = .{ .v = .{ 1, 1 } };
    const v = s.v;
    _ = v;
}

const s2: Strukt = .{ .v = .{ 1, 1 } };
test "fails" {
    const v = s2.v;
    _ = v;
}
```
</details>

```sh
$ uname -a
Linux archlinux 6.1.2-arch1-1 #1 SMP PREEMPT_DYNAMIC Sat, 31 Dec 2022 17:40:35 +0000 x86_64 GNU/Linux

$ zig version
0.11.0-dev.1247+87b223428

$ zig test repro.zig |& cat
1/2 test.passes... OK
2/2 test.fails... OK
All 2 tests passed.
```
